### ISSUE TITLE
Wrong Podman-CLI version detected on startup

### ISSUE THREAD

**Issue by Danik1601** [iongion/podman-desktop-companion#84]
Every time the app starts it detects the installed Podman version wrong until tested manually. Same things happen after each restart of the app. For more information see screenshot and video below.
Edit 1: P.S. as you may see in the video, version number in the bottom left corner does not update at all even after testing manually.
Edit 2: Strange thing happens if I switch to Docker. Upon launch it show that version 20.10.14 is installed, but I have no Docker installed at all. After manual testing it shows "Program is available" instead of version number.

Fedora Silverblue 36.20220515.0
Podman 4.1.0
Podman Desktop Companion 4.1.0-rc.26 (x86_64) (Flatpak)

![Screenshot from 2022-05-15 22-05-35](https://user-images.githubusercontent.com/17238959/168489890-d98e071a-a02f-49b9-932b-d19ca335c3e1.png)

https://user-images.githubusercontent.com/17238959/168490049-6ca81f85-ba96-4bd1-8b34-4d0f5170dead.mp4

**Comment by iongion** [comment]
Not the prettiest thing, will address in next releases soon. That is the baseline version I use to develop the app, but it should be replaced properly as detection happens every time the app is started.

I would like a more ideal approach here, support as many versions as possible, toggle features on/off depending on what is found.

Thank you for reporting, it is very cool I see it at work, what OS do you have ? just as curiosity.

**Comment by Danik1601** [comment]
> will address in next releases soon

Cool, thanks!

> what OS do you have?

Not sure what you meant exactly since I already included the OS in the report so I will just give you all the information.
This is Fedora 36 Silverblue Edition with Dash to Panel GNOME extension running in the Oracle VirtualBox on Windows 11 machine.

By the way thank you for the app. What I see is very promising. Looks like you have made it in a few month with significant updates over very short periods of time. I have no idea what ninja level coding technique you are using but that is very cool. Looking forward to use the app on regular basis after it will make it's way on Flathub.

**Comment by iongion** [comment]
Hehe, thanks, passion man :D (Sorry, I meant desktop environment / manager ... GNOME :D)

**Comment by iongion** [comment]
Issued an evaluation version for you to test, let me know, grateful for your words and encouragements.

https://mega.nz/file/y1E1wRoJ#ym63rehj7zJimw3p1XChvPjPd0GEQ39WZ-m3eNyEgZM

```
md5sum src/app/dist/podman-desktop-companion-x86_64-4.1.0-rc.27.flatpak 
462fef83ce3dd0c5cf7c0300f47ddb0e  src/app/dist/podman-desktop-companion-x86_64-4.1.0-rc.27.flatpak
```

**Comment by Danik1601** [comment]
Works like clockwork now!
But... I noticed that Docker version is still shown as 20.10.14 on startup until "test" button is pressed on my Fedora Silverblue VM (with no Docker installed of course). So I created another VM, this time with Fedora Workstation 36, installed docker on it and launched the program. Everything worked great, version 20.10.16 (latest by now, installed by me) detected upon launch without manual testing.
The next thing I decided to do is to completely remove Docker and Podman from my Fedora installation to see if the same thing happens to Podman as well. Turns out that the same thing happens to Podman version detection. If there is no Podman installed, the program shows that version 4.0.3 is installed (until tested manually).

https://user-images.githubusercontent.com/17238959/168705898-4494b0d0-e324-41a6-99d6-1e251c25e3d0.mp4

The other thing I noticed is that the program checks for installed version repeatedly after successful connection (at least such animation is played). Sorry for bad video quality, GNOME recorder tool couldn't capture it for some reason.

https://user-images.githubusercontent.com/17238959/168707759-92d9a2fc-82be-426d-aadd-23c29e6f01c6.mp4

**Comment by iongion** [comment]
Awesome, you are right, still the baseline popping in, will take care of it soon.
The UI now receives this as source of truth for each connection

- `expected` - meaning baseline
- `user` - user overrides (what you type)
- `detected` - runing `which` / `where` / `where-is` to find followed by `podman --version` or `docker --version`
- `current` - the merging of the above in that order `expected`, `user`, `detected`

Now if `detected` and `user` are missing, I could realize and avoid displaying the `expected`

![image](https://user-images.githubusercontent.com/867609/168733111-1ceb077a-d210-48cc-a593-45c2082adcbc.png)

**Comment by iongion** [comment]
Awesome, you are right, still the baseline popping in, will take care of it soon.
The UI now receives this as source of truth for each connection

- `expected` - meaning baseline
- `user` - user overrides (what you type)
- `detected` - runing `which` / `where` / `where-is` to find followed by `podman --version` or `docker --version`
- `current` - the merging of the above in that order `expected`, `user`, `detected`

Now if `detected` and `user` are missing, I could realize and avoid displaying the `expected`

![image](https://user-images.githubusercontent.com/867609/168733111-1ceb077a-d210-48cc-a593-45c2082adcbc.png)

**Comment by iongion** [comment]
Done, all should be addressed in this release

https://github.com/iongion/podman-desktop-companion/releases/tag/4.1.0-rc.28

**Comment by Danik1601** [comment]
Installed 4.1.0-rc.28 (Flatpak). Everything working as expected with "Not present in path" message for both Podman and Docker in case of their absence from the machine. Million thanks!

I also decided to test Windows version of rc.28 (because why not?) and discovered that the problem is partially persistent on Windows. It is working correctly for "virtualized" section, but in "Custom WSL" section the app shows version 3.4.2 for Podman and version 20.10.14 for Docker even after manual testing even thou "Program could not be reached" message is displayed in top right corner. I can not test macOS version since I don't have the hardware, but I will try to tinker with VM once I have some free time and report it back to you if I succeed.

![image](https://user-images.githubusercontent.com/17238959/168802936-1c8bf6bb-c10f-475c-97ea-792f667ff930.png)

![image](https://user-images.githubusercontent.com/17238959/168802946-dbb818d4-8344-44e8-9e7f-5c3ae41c6f73.png)

**Comment by iongion** [comment]
I am beyond grateful for your help, I just issued a non released `4.1.0-rc.29` - I think it should do it will all WSL situation

> Recommended - Although podman cli is compatible with docker cli, it would be appreciated if you don't cheat :sweat_smile:  and really set the path to podman while on `podman` connections `tab`

![image](https://user-images.githubusercontent.com/867609/168898130-4557c80c-a77a-4ac9-8564-fc48267caf5c.png)

https://mega.nz/file/mwM3RLRS#9n5ZSpB0Tf3OEhiUK0ZADbz6ehh7h7hbK_ZQIe-Vito

```
md5sum podman-desktop-companion-x64-4.1.0-rc.29.exe                                                                                                                                              
50870bfeeefd5979413acb591b957229  podman-desktop-companion-x64-4.1.0-rc.29.exe
```

Once you give the green flag, it goes to prod (website links update)

**Comment by Danik1601** [comment]
4.1.0-rc.29. WSL version detection still acts the same - broken on startup :-(

https://user-images.githubusercontent.com/17238959/168937101-38d965af-e88f-487c-bff8-7b53e7dc91b0.mp4

P.S. Working on macOS VM

**Comment by iongion** [comment]
I think I might understand it better, when you switch the WSL distro, can you press the button on the right to it `Click to trigger automatic detection` and see what happens ?

![image](https://user-images.githubusercontent.com/867609/168957267-f3f72c88-30a0-47d5-ae32-627e180b3933.png)

**Comment by Danik1601** [comment]
Sorry it took me some time - I completely nuked my Hyper-V and WSL in order to install macOS, so I had to switch to another machine for some more WSL testing.
During testing I discovered even more strange behavior. Everything is shown in the videos below.

So... here's what's happening:
In the first video there is no Podman installed in Windows environment. The program correctly detects that, but still fails to correctly detect version installed in WSL on startup. Everything else is in the video - I just decided to push all the buttons for you to see what's happening, all the messages, etc, etc.
But in the second video you can see more interesting stuff happening. I installed Podman 4.1.0 for Windows in between the videos. First of all, in "Virtualized" section there is "Podman machine required" and "Version 4.1.0 detected" shown at the same time. Then, Podman for Windows version is detected correctly on startup, but for some reason this version is saved and shown when I switch to WSL installation.

Sorry if expressed it in a complicated or a twisted way - English is not my native language. Please refer to the videos and software specifications listed below if there is something unclear for you.

First video:
- Windows 11 21H2
- Podman Desktop Companion 4.1.0-rc.29
- Docker Desktop 4.8.2 (installed in Windows)
- Docker Engine 2.10.14 (installed in Windows)
- WSL 2
- Ubuntu 22.04 Jammy Jellyfish (installed in WSL)
- Podman 3.4.4 (installed in Ubuntu WSL)
- Docker 2.10.16 (installed in Ubuntu WSL)

Second video:
- Everything listed above
- Podman 4.1.0 for Windows 

https://user-images.githubusercontent.com/17238959/169389896-8229cc23-64ac-4667-aab7-afdfec7b4f14.mp4

https://user-images.githubusercontent.com/17238959/169389927-3f872e45-ae1e-4532-b4d4-77256c3ce8b1.mp4

P.S. macOS VM is on the way. It is ready and LIMA is installed on it. I'm just figuring out how to deal with LIMA since I have zero experience with it

**Comment by Danik1601** [comment]
Sorry, seems like no LIMA testing by me.
I installed macOS 10.15 Catalina in VirtualBox using [this script](https://github.com/myspaghetti/macos-virtualbox) but currently [unable to use nested virtualization](https://github.com/myspaghetti/macos-virtualbox/issues/604) and, as a result, LIMA.

I have managed to only test "native" Podman and Docker installations.

First video
- macOS 10.15.7 Catalina (with latest security patch)
- Podman Desktop Companion 4.1.0-rc.28.p

Second video
- Everything listed above
- Podman 4.1.0 (installed using Homebrew)
- Docker 2.10.16 (installed using Homebrew)

https://user-images.githubusercontent.com/17238959/169656402-4864b0bf-db6a-4cb7-b02f-5aab8171e1f9.mp4


https://user-images.githubusercontent.com/17238959/169656404-c29764e8-2ee7-4567-a9e3-faca4c7584b8.mp4
